### PR TITLE
Support windows

### LINKED ISSUE #72: [Cross Platform]Support Windows
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/async_simple/issues) and found no similar issues.

### What happened + What you expected to happen
We have supported Mac, the next is Windows, at least make sure supporting Windows when using c++20 coroutine.

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by qicosmos**
## Why

Closes #72 

## What is changing

name conflicts to windows:
- _try --> _try_value
- Yield() --> Yield_t()

not exist functions on windows:
- ```__builtin_expect --> #define __builtin_expect(EXP, C) (EXP)```
- ```__attribute__((__always_inline__)) --> #define FL_INLINE inline```

change some compile options for windows:
- pthread
- async_simple static lib
- some c++ compile options

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -67,24 +76,27 @@ set(HEADERS_PATH
...
+        list(APPEND deplibs "-m32")
+    else()
+        message("-- Use flag -m64")
+        list(APPEND CXX_FLAGS "-m64")
+        list(APPEND CXX_FLAGS "-DTARGET_64")
+        #list(APPEND deplibs "-m64")
+    endif()
 endif()
 
 if (CMAKE_BUILD_TYPE STREQUAL "Debug")
```
Comment (ChuanqiXu9) [inline CMakeLists.txt:102]: I guess `-fsanitize=address` is not meaningful in windows.

File: async_simple/CMakeLists.txt [hunk 2]
```diff
@@ -30,15 +30,15 @@ if(UTHREAD)
   list(APPEND SRCS ${uthread_asm_src})
 endif()
 
-add_library(async_simple_static STATIC ${SRCS})
-add_library(async_simple SHARED ${SRCS})
+add_library(async_simple STATIC ${SRCS})
```
Comment (ChuanqiXu9) [inline async_simple/CMakeLists.txt:33]: Is this change necessary?

File: async_simple/Common.h [hunk 3]
```diff
@@ -18,10 +18,18 @@
 
 #include <stdexcept>
 
+#ifdef _WIN32
+#define __builtin_expect(EXP, C) (EXP)
+#endif
+
 #define FL_LIKELY(x) __builtin_expect((x), 1)
 #define FL_UNLIKELY(x) __builtin_expect((x), 0)
```
Comment (ChuanqiXu9) [inline async_simple/Common.h:26]: We could use `[[likely]]` one day.

File: async_simple/Common.h [hunk 4]
```diff
@@ -18,10 +18,18 @@
...
 
+#ifdef _WIN32
+#define FL_INLINE inline
```
Comment (ChuanqiXu9) [inline async_simple/Common.h:29]: Better rename `FL_INLINE` to `AS_INLNIE`

File: async_simple/FutureState.h [hunk 5]
```diff
@@ -168,8 +168,8 @@ class FutureState {
     }
 
 public:
-    Try<T>& getTry() noexcept { return _try; }
-    const Try<T>& getTry() const noexcept { return _try; }
+    Try<T>& getTry() noexcept { return _try_value; }
```
Comment (ChuanqiXu9) [inline async_simple/FutureState.h:171]: Why the renaming is necessary?

File: async_simple/coro/Lazy.h [hunk 6]
```diff
@@ -43,7 +43,7 @@ class Lazy;
 // ```
 //
 // This would suspend the executing coruotine.
-struct Yield {};
+struct Yield_t {};
```
Comment (ChuanqiXu9) [inline async_simple/coro/Lazy.h:46]: Is this necessary? Why?

File: async_simple/experimental/coroutine.h [hunk 7]
```diff
@@ -55,7 +55,8 @@ template <class P> struct hash<coroutine_handle<P>>;
 // we could only use self-provided coroutine header.
 // Note: the <coroutine> header in libc++ is available for both clang and gcc.
 // And the outdated <experimental/coroutine> is available for clang only.
-#if (__cplusplus <= 201703L) || (defined(__clang__) && defined(__GLIBCXX__))
+#if (__cplusplus <= 201703L && !defined(_MSC_VER)) || \
```
Comment (ChuanqiXu9) [inline async_simple/experimental/coroutine.h:58]: Edit the comment too.

File: async_simple/executors/SimpleIOExecutor.h [hunk 8]
```diff
@@ -20,7 +20,9 @@
 #ifndef ASYNC_SIMPLE_HAS_NOT_AIO
 #include <libaio.h>
 #endif
+#ifndef _WIN32
 #include <sys/syscall.h> /* For SYS_xxx definitions */
```
Comment (RainMark) [inline async_simple/executors/SimpleIOExecutor.h:24]: this header can be removed directly？

**Comment by qicosmos** [comment]
I will have a try and show more error messages after testing on Windows later.

File: async_simple/coro/Lazy.h (hunk 6 above)
Comment (qicosmos) [reply async_simple/coro/Lazy.h]: Has reverted, it's related with asio.hpp header sequence.

File: async_simple/FutureState.h (hunk 5 above)
Comment (qicosmos) [reply async_simple/FutureState.h:171]: Can't use _try or __try variable in windows, will cause confilict:https://godbolt.org/z/KcYMsTvx5

File: async_simple/Common.h (hunk 3 above)
Comment (qicosmos) [reply async_simple/Common.h]: Actually msvc, gcc and clang12+ supported likely/unlikely, but ci clang version is clang11, will have a unknown-attributes warning. Maybe add a compile flag to avoid this warning errorr.

File: async_simple/coro/Lazy.h (hunk 6 above)
Comment (qicosmos) [reply async_simple/coro/Lazy.h]: Can be removed.

File: async_simple/executors/SimpleIOExecutor.h (hunk 8 above)
Comment (qicosmos) [reply async_simple/executors/SimpleIOExecutor.h]: Useless, can be removed.

File: async_simple/CMakeLists.txt (hunk 2 above)
Comment (qicosmos) [reply async_simple/CMakeLists.txt:33]: If use the original config, will generate async_imple.dll on windows, and all the test cases depends on the async_simple.lib and compile error on windows, so here don't use set_target_properties. Explicit generate static and shared lib.

File: async_simple/Common.h [hunk 9]
```diff
@@ -18,10 +18,15 @@
 
 #include <stdexcept>
 
-#define FL_LIKELY(x) __builtin_expect((x), 1)
-#define FL_UNLIKELY(x) __builtin_expect((x), 0)
+#ifdef _WIN32
+#define __builtin_expect(EXP, C) (EXP)
+#endif
```
Comment (ChuanqiXu9) [inline async_simple/Common.h:23]: Is this still useful?

File: async_simple/Common.h [hunk 10]
```diff
@@ -18,10 +18,15 @@
...
 
-#define FL_INLINE __attribute__((__always_inline__)) inline
+#ifdef _WIN32
+#define AS_INLINE inline
```
Comment (ChuanqiXu9) [inline async_simple/Common.h:26]: `inline` should have other meaning than suggesting a inlining. Let's define AS_INLINE as empty for windows.

File: async_simple/CMakeLists.txt (hunk 2 above)
Comment (ChuanqiXu9) [reply async_simple/CMakeLists.txt:33]: Now the behavior changes more. Currently, the shared library wouldn't get installed. Also the name changes. Both is not good. I would suggest to wrap line 38 for windows only.

File: async_simple/Common.h [hunk 11]
```diff
@@ -31,9 +36,14 @@ namespace async_simple {
...
 inline void logicAssert(bool x, const char* errorMsg) {
-    if (FL_LIKELY(x)) {
+    if (x)
+#if defined(__clang__) && __clang_major__ < 12
+    {
         return;
     }
+#else
+        [[likely]] { return; }
+#endif
```
Comment (ChuanqiXu9) [inline async_simple/Common.h:46]: This looks redundant. I would suggest:
```suggestion
    if (x) AS_LIKELY
        return;
```

File: async_simple/executors/test/CMakeLists.txt [hunk 12]
```diff
@@ -1,7 +1,11 @@
 file(GLOB executor_test_src "*.cpp")
 add_executable(async_simple_executor_test ${executor_test_src} ${PROJECT_SOURCE_DIR}/async_simple/test/dotest.cpp)
 
-target_link_libraries(async_simple_executor_test async_simple -pthread ${deplibs} ${testdeplibs})
+if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+    target_link_libraries(async_simple_executor_test async_simple ${deplibs} ${testdeplibs})
+else()
+    target_link_libraries(async_simple_executor_test async_simple -pthread ${deplibs} ${testdeplibs})
+endif()
```
Comment (ChuanqiXu9) [inline async_simple/executors/test/CMakeLists.txt:8]: From the context, it looks like we could use ${CMAKE_THREAD_LIBS_INIT} here?

**Review by ChuanqiXu9 — CHANGES_REQUESTED** [review]

File: async_simple/executors/test/CMakeLists.txt (hunk 12 above)
Comment (qicosmos) [reply async_simple/executors/test/CMakeLists.txt]: No need change, revert is ok on windows, maybe i did not compile all projects cause such error before.

File: async_simple/Common.h (hunk 11 above)
Comment (qicosmos) [reply async_simple/Common.h]: Good

File: async_simple/Common.h (hunk 10 above)
Comment (qicosmos) [reply async_simple/Common.h]: OK

File: async_simple/Common.h (hunk 9 above)
Comment (qicosmos) [reply async_simple/Common.h]: Removed

**Review by ChuanqiXu9 — APPROVED** [review]
The formatted code for `LIKELY` macro looks ugly. But it looks like we couldn't do any thing better. We need update documents and CI for windows too. Let's do it in another PR.
